# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1251236460 14400
#      Tue Aug 25 17:41:00 2009 -0400
# Node ID c60a9e1a03727db13a5a8670dd553b7b5f7d487c
# Parent  d280bfa04996fa3bf2e1d02f22dea56fdb6d6d62
try to avoid gnuplot zombies

diff --git a/scripts/ChangeLog b/scripts/ChangeLog
--- a/scripts/ChangeLog
+++ b/scripts/ChangeLog
@@ -1,8 +1,15 @@
+2009-08-25  John W. Eaton  <jwe@octave.org>
+
+	* plot/__gnuplot_open_stream__.m: Save pid in __plot_stream__ property.
+	* plot/gnuplot_drawnow.m: Wait for gnuplot subprocess when printing.
+	From Ben Abbott <bpabbott@mac.com>, Rob Mahurin <rob@utk.edu>, and
+	Dmitri Sergatskov <dasergatskov@gmail.com>.
+
 2009-08-22  David Bateman  <dbateman@free.fr>
 
 	* plot/__add_datasource__.m: Correct test for "datasource" argument
 	* plot/__countour__.m: Add edgecolor properties and make it an alias
 	for linecolor with the value "auto" being "flat" for the edgecolor.
 
 2009-08-17  John W. Eaton  <jwe@octave.org>
 
diff --git a/scripts/plot/__gnuplot_open_stream__.m b/scripts/plot/__gnuplot_open_stream__.m
--- a/scripts/plot/__gnuplot_open_stream__.m
+++ b/scripts/plot/__gnuplot_open_stream__.m
@@ -23,16 +23,18 @@
 ## Created: 2009-04-11
 
 function plot_stream = __gnuplot_open_stream__ (npipes, h)
   [prog, args] = gnuplot_binary ();
   if (npipes > 1)
     [plot_stream(1), plot_stream(2), pid] = popen2 (prog, args{:});
     if (pid < 0)
       error ("__gnuplot_open_stream__: failed to open connection to gnuplot.");
+    else
+      plot_stream(3) = pid;
     endif
   else
     plot_stream = popen (sprintf ("%s ", prog, args{:}), "w");
     if (plot_stream < 0)
       error ("__gnuplot_open_stream__: failed to open connection to gnuplot.");
     endif
   endif
   if (nargin > 1)
diff --git a/scripts/plot/gnuplot_drawnow.m b/scripts/plot/gnuplot_drawnow.m
--- a/scripts/plot/gnuplot_drawnow.m
+++ b/scripts/plot/gnuplot_drawnow.m
@@ -62,19 +62,22 @@ function gnuplot_drawnow (h, term, file,
       else
         error ("gnuplot_drawnow: the gnuplot terminal, \"%s\", is not available.",
                gnuplot_trim_term (term))
       endif
     unwind_protect_cleanup
       set (h, "__plot_stream__", default_plot_stream);
       if (! isempty (plot_stream))
         pclose (plot_stream(1));
-        if (numel (plot_stream) == 2)
+        if (numel (plot_stream) > 1)
           pclose (plot_stream(2));
         endif
+	if (numel (plot_stream) > 2)
+	  waitpid (plot_stream(3));
+	endif
       endif
       if (! isempty (fid))
         fclose (fid);
       endif
     end_unwind_protect
   elseif (nargin == 1)
     ##  Graphics terminal for display.
     plot_stream = get (h, "__plot_stream__");
diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,12 @@
+2009-08-25  John W. Eaton  <jwe@octave.org>
+
+	* graphics.cc (gnuplot_backend::send_quit): Wait for gnuplot process.
+
 2009-08-24  Jaroslav Hajek  <highegg@gmail.com>
 
 	* OPERATORS/op-m-m.cc: Install .*= and ./= operators.
 	* OPERATORS/op-fm-fm.cc: Ditto.
 	* OPERATORS/op-cm-cm.cc: Ditto.
 	* OPERATORS/op-fcm-fcm.cc: Ditto.
 	* OPERATORS/op-int.h: Ditto.
 
diff --git a/src/graphics.cc b/src/graphics.cc
--- a/src/graphics.cc
+++ b/src/graphics.cc
@@ -2037,33 +2037,41 @@ public:
     { return 72.0; }
 
   Matrix get_screen_size (void) const
     { return Matrix (1, 2, 0.0); }
 
 private:
   void send_quit (const octave_value& pstream) const
     {
-      if (! pstream.is_empty())
+      if (! pstream.is_empty ())
 	{
 	  octave_value_list args;
 	  Matrix fids = pstream.matrix_value ();
 
 	  if (! error_state)
 	    {
 	      args(1) = "\nquit;\n";
-	      args(0) = octave_value (fids (0));
+	      args(0) = fids(0);
 	      feval ("fputs", args);
+
 	      args.resize (1);
 	      feval ("fflush", args);
 	      feval ("pclose", args);
+
 	      if (fids.numel () > 1)
 		{
-		  args(0) = octave_value (fids (1));
+		  args(0) = fids(1);
 		  feval ("pclose", args);
+
+		  if (fids.numel () > 2)
+		    {
+		      args(0) = fids(2);
+		      feval ("waitpid", args);
+		    }
 		}
 	    }
 	}
     }
 };
 
 graphics_backend
 graphics_backend::default_backend (void)
